policies/recipes/artifacts_test.go

```diff
@@ -0,0 +1,302 @@
+			Content:    []byte("inside the file"),
+		},
+	})
+	defer server.Stop()
+
+	getGCSClient = func(ctx context.Context) (*storage.Client, error) {
+		return nil, errors.New("Error creating storage client")
+	}
+
+	gcsartifact := osconfigpb.SoftwareRecipe_Artifact_Gcs{
+		Object: "some-object.txt",
+		Bucket: "some-bucket",
+	}
+	artifacts := []*osconfigpb.SoftwareRecipe_Artifact{
+		&osconfigpb.SoftwareRecipe_Artifact{
+			Id:            "id1",
+			Artifact:      &osconfigpb.SoftwareRecipe_Artifact_Gcs_{Gcs: &gcsartifact},
+			AllowInsecure: true,
+		},
+	}
+	art, err := FetchArtifacts(context.Background(), artifacts, "/tmp/")
+
+	if len(art) != 0 {
+		t.Errorf("unexpted artifact fetched")
+	}
+
+	if err == nil {
+		t.Errorf("Expected (client error), got(%+v)\n", err)
+	}
+}
+
+func TestFetchArtifacts_http_happycase(t *testing.T) {
+	// Start a local HTTP server
+	server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		// Test request parameters
+		rw.Write([]byte("ok"))
+		// Send response to be tested
+	}))
+	// Close the server when test finishes
+	defer server.Close()
+
+	getHTTPClient = func() (*http.Client, error) {
+		return server.Client(), nil
+	}
+	remoteartifact := osconfigpb.SoftwareRecipe_Artifact_Remote{
+		Uri:      fmt.Sprintf("%s/agent.deb", server.URL),
+		Checksum: "",
+	}
+	artifacts := []*osconfigpb.SoftwareRecipe_Artifact{
+		&osconfigpb.SoftwareRecipe_Artifact{
+			Id:            "id1",
+			Artifact:      &osconfigpb.SoftwareRecipe_Artifact_Remote_{Remote: &remoteartifact},
+			AllowInsecure: true,
+		},
+	}
+
+	art, err := FetchArtifacts(context.Background(), artifacts, "/tmp/")
+
+	if !exists("/tmp/id1.deb") {
+		t.Errorf("expected fetched artifacts")
+	}
+
+	if err = os.Remove("/tmp/id1.deb"); err != nil {
+		t.Fatalf("could not clean up test data")
+	}
+
+	if art["id1"] != "/tmp/id1.deb" {
+		t.Errorf("expected file entry, but not present")
+	}
+}
+
+func TestFetchArtifacts_http_InvalidURL(t *testing.T) {
+	// Start a local HTTP server
+	server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		// Test request parameters
+		rw.Write([]byte("ok"))
+		// Send response to be tested
+	}))
+	// Close the server when test finishes
+	defer server.Close()
+
+	getHTTPClient = func() (*http.Client, error) {
+		return server.Client(), nil
+	}
+	remoteartifact := osconfigpb.SoftwareRecipe_Artifact_Remote{
+		Uri:      fmt.Sprintf("%sagent.deb", server.URL),
+		Checksum: "",
+	}
+	artifacts := []*osconfigpb.SoftwareRecipe_Artifact{
+		&osconfigpb.SoftwareRecipe_Artifact{
+			Id:            "id1",
+			Artifact:      &osconfigpb.SoftwareRecipe_Artifact_Remote_{Remote: &remoteartifact},
+			AllowInsecure: true,
+		},
+	}
+
+	art, err := FetchArtifacts(context.Background(), artifacts, "/tmp/")
+
+	if len(art) != 0 {
+		t.Errorf("unexpected artifact downloaded")
+	}
+
+	if err == nil || !strings.Contains(err.Error(), "Could not parse url") {
```

Comment:
it's fragile to check for this specific string - is the nil check not enough?

Reply:
this test is re-written because of refactoring.
however; i am using the string check for errors in other places as well. nil check is just not enough since it can be too broad.
generally; to make assertion on errors; it is a common practice to introduce new error objects (object oriented programming) and then cast the error recieved from downstream to one of the appropriate one's which is introduced in their application.
we can do something similar in our application; but it would still be a string check.

Reply:
i'm familiar with named errors, but i don't think this practice is very common in go development

Reply:
yeah.. even i could not find examples of named errors.